Update aws-c* submodules

vitlibar opened the pull request:
### Changelog category:
- Not for changelog

Update `aws-sdk-cpp` (contrib/aws) from [1.8.121](https://github.com/aws/aws-sdk-cpp/tree/1.8.121) to [1.10.36](https://github.com/aws/aws-sdk-cpp/tree/1.10.36).

Also update dependency modules: `aws-c-auth`, `aws-c-cal`, `aws-c-common`, `aws-c-compression`, `aws-c-event-stream`,
`aws-c-http`, `aws-c-io`, `aws-c-mqtt`, `aws-c-s3`, `aws-c-sdkutils`, `aws-checksums`, `aws-crt-cpp`, `aws-s2n-tls`
to the versions corresponding to `aws-sdk-cpp 1.10.36`

This closes #42221.
This closes #43011.
This closes #37431.
This closes #32439.

vitlibar committed:
Update AWS SDK to version 1.10.36

vitlibar committed:
Disable loading EC2-metadata.

vitlibar committed:
Fix checking for embedded error in PocoHTTPClient.

vitlibar committed:
Fix finding pthread_atfork() while linking ppc64le build.

vitlibar committed:
Disable memory sanitizer for events from epoll

vitlibar committed:
Fix race condition in cJSON_Parse.* calls.

vitlibar committed:
Fix support for boringssl.

vitlibar committed:
Correctly handle that dlopen(NULL) returns NULL in ClickHouse.

vitlibar committed:
Remove subsubmodules contrib/aws/crt and contrib/aws-crt-cpp/crt
because they were moved to the contrib folder.

alesapin commented on .gitmodules:
```diff
@@ -294,3 +294,33 @@
 	url = https://github.com/ridiculousfish/libdivide.git
+[submodule "contrib/aws-crt-cpp"]
+	path = contrib/aws-crt-cpp
+	url = https://github.com/ClickHouse/aws-crt-cpp.git
```
Why half of the modules are from `awslabs` and a half from `ClickHouse`?

vitlibar replied:
For each module It depends on whether we've changed that module or not.

alesapin commented on contrib/CMakeLists.txt:
```diff
@@ -115,12 +115,25 @@ endif()
+add_contrib (aws-cmake
     aws
+    aws-c-auth
+    aws-c-cal
```
Why do we need so many modules now? How has it worked before without them?

vitlibar replied:
It worked before it was before AWS SDK team added their own submodule system, see
https://github.com/aws/aws-sdk-cpp/commit/bb964685686c211d078fba7b8c840b73e16430de

alesapin commented on src/IO/S3/PocoHTTPClient.cpp:
```diff
@@ -177,7 +177,7 @@ namespace
     {
         auto query_params = request.GetQueryStringParameters();
-        if (request.HasHeader("z-amz-copy-source"))
+        if (request.HasHeader("x-amz-copy-source"))
```
??? Did they change the format of the header, or what?

vitlibar replied:
It's always been `x-amz-copy-source` since at least 2015. See
https://github.com/aws/aws-sdk-cpp/blame/main/aws-cpp-sdk-s3/source/model/CopyObjectRequest.cpp#L176
and
https://github.com/aws/aws-sdk-cpp/blob/5fb5947c201b2fa89eab3f64142f59f0251704ef/aws-cpp-sdk-s3/source/model/CopyObjectRequest.cpp#L108

alesapin commented on src/IO/S3Common.cpp:
```diff
@@ -677,6 +677,12 @@ namespace S3
     {
         aws_options = Aws::SDKOptions{};
+
+        /// We don't load EC2 metadata.
```
Why? It's one of the most popular ways to get auth info. Some cloud providers for clickhouse use it.

vitlibar replied:
Here I'm just keeping EC2 metadata disabled as it was before my PR. See https://github.com/ClickHouse/aws-sdk-cpp/commit/4bd1ca4af129d22ea7d53b8148520e08bd77de32. I just slightly changed the way how it's implemented in the code.

Why it was disabled by us in https://github.com/ClickHouse/aws-sdk-cpp/commit/4bd1ca4af129d22ea7d53b8148520e08bd77de32:
To create `S3::PocoHTTPClientConfiguration` we use `S3::ClientFactory::instance().createClientConfiguration()` and `S3::ClientFactory::instance()` calls `S3::ClientFactory::ClientFactory()` which calls `Aws::InitAPI()` which calls `Aws::Internal::InitEC2MetadataClient()`.

And  `Aws::Internal::InitEC2MetadataClient()` uses `Aws::Client::ClientConfiguration` instead of our subclass `S3::PocoHTTPClientConfiguration`. However our function `PocoHTTPClientFactory::CreateHttpClient()` assumes that the passed `clientConfiguration` is a `S3::PocoHTTPClientConfiguration` and crashes.

To fix that we could pass our `S3::PocoHTTPClientConfiguration` to `Aws::InitAPI()` and then to `Aws::Internal::InitEC2MetadataClient()`, but if we do so we will have circular dependency `S3::PocoHTTPClientConfiguration` -> `S3::ClientFactory` -> `Aws::InitAPI()` -> `S3::PocoClientConfiguration`, and it's not very trivial how to fix that.

Anyway I'm not doing things worse here as I said. I suppose we can try to resolve that circular dependency in a separate PR.

vitlibar replied:
I'll investigate that further.

vitlibar replied:
We load EC2 metadata by ourselves so it should be disabled in aws sdk core, but not by changing that environment variable. I changed the code and wrote a comment about it - see https://github.com/ClickHouse/ClickHouse/pull/43020/commits/9014be89681c4509485e7e809fdda27ccef817d6

vitlibar committed:
Fix disable loading EC2 metadata by aws-sdk-core.

vitlibar committed:
Enabling override of endpoint used to connect for EC2 Metadata Service.
This commit applies changes https://github.com/aws/aws-sdk-cpp/commit/742b371 and
https://github.com/aws/aws-sdk-cpp/commit/34e73e9 to our fork of Aws::Internal::EC2MetadataClient.

vitlibar committed:
Add test for getting credentials from AWS EC2 metadata.

vitlibar commented:
@alesapin 
Now everything is alright, I even added a test checking that EC2 metadata works.

alesapin commented on src/IO/S3Common.cpp:
```diff
@@ -286,12 +285,50 @@ class AWSEC2MetadataClient : public Aws::Internal::AWSHttpResourceClient
+        {
+            if (ec2_metadata_service_endpoint_mode.length() == 4)
+            {
+                if (Aws::Utils::StringUtils::CaselessCompare(ec2_metadata_service_endpoint_mode.c_str(), "ipv4"))
```
Looks slightly unreliable :)

azat commented:
This upgrade introduces memory leaks, see https://s3.amazonaws.com/clickhouse-test-reports/40003/723dad3610d1ba8643f634d255becd5b9d2008fb/integration_tests__asan__[2/6].html:

```
Direct leak of 152 byte(s) in 1 object(s) allocated from:
    #0 0xe958dee in malloc (/usr/bin/clickhouse+0xe958dee) (BuildId: 5bb74c5832b8b48b64e089f56daaea78ac0c8d42)
    #1 0x418c9873 in OPENSSL_malloc build_docker/../contrib/boringssl/crypto/mem.c:139:15
    #2 0x417acaaf in EVP_CIPHER_CTX_new build_docker/../contrib/boringssl/crypto/fipsmodule/cipher/cipher.c:76:25
    #3 0x3dadc892 in s2n_drbg_instantiate (/usr/bin/clickhouse+0x3dadc892) (BuildId: 5bb74c5832b8b48b64e089f56daaea78ac0c8d42)
    #4 0x3dadb2db in s2n_ensure_initialized_drbgs s2n_random.c
    #5 0x3dadb052 in s2n_rand_init (/usr/bin/clickhouse+0x3dadb052) (BuildId: 5bb74c5832b8b48b64e089f56daaea78ac0c8d42)
    #6 0x3dad4ef0 in s2n_init (/usr/bin/clickhouse+0x3dad4ef0) (BuildId: 5bb74c5832b8b48b64e089f56daaea78ac0c8d42)
    #7 0x3d990cdc in aws_tls_init_static_state (/usr/bin/clickhouse+0x3d990cdc) (BuildId: 5bb74c5832b8b48b64e089f56daaea78ac0c8d42)
    #8 0x3db8ca3a in aws_mqtt_library_init (/usr/bin/clickhouse+0x3db8ca3a) (BuildId: 5bb74c5832b8b48b64e089f56daaea78ac0c8d42)
    #9 0x3dae97b9 in Aws::Crt::ApiHandle::ApiHandle(aws_allocator*) (/usr/bin/clickhouse+0x3dae97b9) (BuildId: 5bb74c5832b8b48b64e089f56daaea78ac0c8d42)
    #10 0x3d1f91eb in Aws::InitializeCrt() (/usr/bin/clickhouse+0x3d1f91eb) (BuildId: 5bb74c5832b8b48b64e089f56daaea78ac0c8d42)
    #11 0x3d1f541a in Aws::InitAPI(Aws::SDKOptions const&) (/usr/bin/clickhouse+0x3d1f541a) (BuildId: 5bb74c5832b8b48b64e089f56daaea78ac0c8d42)
    #12 0x2e55c576 in DB::S3::ClientFactory::ClientFactory() build_docker/../src/IO/S3Common.cpp:717:9
    #13 0x2e55d894 in DB::S3::ClientFactory::instance() build_docker/../src/IO/S3Common.cpp:730:30
    #14 0x30052048 in DB::getClient(Poco::Util::AbstractConfiguration const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context const>, DB::S3ObjectStorageSettings const&) build_docker/../src/Disks/ObjectStorages/S3/diskSettings.cpp:116:60
    #15 0x300401e7 in DB::registerDiskS3(DB::DiskFactory&, bool)::$_0::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, Poco::Util::AbstractConfiguration const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context const>, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IDisk>, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const, std::__1::shared_ptr<DB::IDisk>>>> const&) const build_docker/../src/Disks/ObjectStorages/S3/registerDiskS3.cpp:125:23
```

vitlibar commented:
> This upgrade (likely) introduces memory leaks, see [https://s3.amazonaws.com/clickhouse-test-reports/40003/723dad3610d1ba8643f634d255becd5b9d2008fb/integration_tests__asan__[2/6].html](https://s3.amazonaws.com/clickhouse-test-reports/40003/723dad3610d1ba8643f634d255becd5b9d2008fb/integration_tests__asan__%5B2/6%5D.html):
> 
> ```
> Direct leak of 152 byte(s) in 1 object(s) allocated from:
>     #0 0xe958dee in malloc (/usr/bin/clickhouse+0xe958dee) (BuildId: 5bb74c5832b8b48b64e089f56daaea78ac0c8d42)
>     #1 0x418c9873 in OPENSSL_malloc build_docker/../contrib/boringssl/crypto/mem.c:139:15
>     #2 0x417acaaf in EVP_CIPHER_CTX_new build_docker/../contrib/boringssl/crypto/fipsmodule/cipher/cipher.c:76:25
>     #3 0x3dadc892 in s2n_drbg_instantiate (/usr/bin/clickhouse+0x3dadc892) (BuildId: 5bb74c5832b8b48b64e089f56daaea78ac0c8d42)
>     #4 0x3dadb2db in s2n_ensure_initialized_drbgs s2n_random.c
>     #5 0x3dadb052 in s2n_rand_init (/usr/bin/clickhouse+0x3dadb052) (BuildId: 5bb74c5832b8b48b64e089f56daaea78ac0c8d42)
>     #6 0x3dad4ef0 in s2n_init (/usr/bin/clickhouse+0x3dad4ef0) (BuildId: 5bb74c5832b8b48b64e089f56daaea78ac0c8d42)
>     #7 0x3d990cdc in aws_tls_init_static_state (/usr/bin/clickhouse+0x3d990cdc) (BuildId: 5bb74c5832b8b48b64e089f56daaea78ac0c8d42)
>     #8 0x3db8ca3a in aws_mqtt_library_init (/usr/bin/clickhouse+0x3db8ca3a) (BuildId: 5bb74c5832b8b48b64e089f56daaea78ac0c8d42)
>     #9 0x3dae97b9 in Aws::Crt::ApiHandle::ApiHandle(aws_allocator*) (/usr/bin/clickhouse+0x3dae97b9) (BuildId: 5bb74c5832b8b48b64e089f56daaea78ac0c8d42)
>     #10 0x3d1f91eb in Aws::InitializeCrt() (/usr/bin/clickhouse+0x3d1f91eb) (BuildId: 5bb74c5832b8b48b64e089f56daaea78ac0c8d42)
>     #11 0x3d1f541a in Aws::InitAPI(Aws::SDKOptions const&) (/usr/bin/clickhouse+0x3d1f541a) (BuildId: 5bb74c5832b8b48b64e089f56daaea78ac0c8d42)
>     #12 0x2e55c576 in DB::S3::ClientFactory::ClientFactory() build_docker/../src/IO/S3Common.cpp:717:9
>     #13 0x2e55d894 in DB::S3::ClientFactory::instance() build_docker/../src/IO/S3Common.cpp:730:30
>     #14 0x30052048 in DB::getClient(Poco::Util::AbstractConfiguration const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context const>, DB::S3ObjectStorageSettings const&) build_docker/../src/Disks/ObjectStorages/S3/diskSettings.cpp:116:60
>     #15 0x300401e7 in DB::registerDiskS3(DB::DiskFactory&, bool)::$_0::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, Poco::Util::AbstractConfiguration const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context const>, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::shared_ptr<DB::IDisk>, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const, std::__1::shared_ptr<DB::IDisk>>>> const&) const build_docker/../src/Disks/ObjectStorages/S3/registerDiskS3.cpp:125:23
> ```

The leak will be fixed in https://github.com/ClickHouse/ClickHouse/pull/44942
